internal/certprovider.go

```diff
@@ -1,59 +1,174 @@
+
+//go:embed assets/AWS_NitroEnclaves_Root-G1.sha256.hex
+var AWSNitroEnclavesRootSHA256Hex string
+
+type UnzipAWSRootCertsFunc func(zipBytes []byte) (pemBytes []byte, err error)
+
+func UnzipAWSRootCerts(zipBytes []byte) ([]byte, error) {
+	zipReader, err := zip.NewReader(
+		bytes.NewReader(zipBytes),
+		int64(len(zipBytes)),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("creating zip reader: %w", err)
+	}
+
+	if len(zipReader.File) != 1 {
+		return nil, fmt.Errorf("unexpected file count: %d", len(zipReader.File))
+	}
+	if zipReader.File[0].Name != "root.pem" {
+		return nil, fmt.Errorf("unexpected file name: %s", zipReader.File[0].Name)
+	}
+
+	pemFile, err := zipReader.File[0].Open()
+	if err != nil {
+		return nil, fmt.Errorf("opening zip file: %w", err)
+	}
+
+	pemBytes, err := io.ReadAll(pemFile)
+	if err != nil {
+		return nil, fmt.Errorf("reading pem file: %w", err)
+	}
+
+	return pemBytes, nil
+}
+
+type ExtractRootsFunc func(
+	rootsZIPBytes []byte,
+	rootsDigestHex string,
+	unzipAWSRootCerts UnzipAWSRootCertsFunc,
+) (
+	*x509.CertPool,
+	error,
+)
+
+func ExtractRoots(
+	rootsZIPBytes []byte,
+	rootsDigestHex string,
+	unzipAWSRootCerts UnzipAWSRootCertsFunc,
+) (
+	*x509.CertPool,
+	error,
+) {
+	digest := sha256.Sum256(rootsZIPBytes)
+	digestHex := hex.EncodeToString(digest[:])
+	if digestHex != strings.TrimSpace(rootsDigestHex) {
+		return nil,
+			fmt.Errorf("digest mismatch: %s != %s", digestHex, rootsDigestHex)
+	}
+
+	rootCertsPEM, err := unzipAWSRootCerts(rootsZIPBytes)
+	if err != nil {
+		return nil, fmt.Errorf("unzipping roots: %w", err)
+	}
+
+	roots := x509.NewCertPool()
+	ok := roots.AppendCertsFromPEM(rootCertsPEM)
+	if !ok {
+		return nil, fmt.Errorf("appending cert")
+	}
+	return roots, nil
+}
 
 type NitroCertProvider struct {
-	certs *x509.CertPool
+	RootCerts    *x509.CertPool
+	ExtractRoots ExtractRootsFunc
 }
 
 func NewNitroCertProvider() *NitroCertProvider {
-	return &NitroCertProvider{}
+	return &NitroCertProvider{
+		RootCerts:    nil,
+		ExtractRoots: ExtractRoots,
+	}
 }
 
 func (cp *NitroCertProvider) Roots() (*x509.CertPool, error) {
-	if cp.certs == nil {
-		certs, err := cp.RootsWithCerts(NitroCertPEM)
+	if cp.RootCerts == nil {
+		certs, err := cp.ExtractRoots(
+			AWSNitroEnclavesRootZip,
+			AWSNitroEnclavesRootSHA256Hex,
+			UnzipAWSRootCerts,
+		)
 		if err != nil {
-			return nil, err
+			return nil, fmt.Errorf("extracting roots: %w", err)
 		}
-		cp.certs = certs
+		cp.RootCerts = certs
 	}
 
-	return cp.certs, nil
+	return cp.RootCerts, nil
 }
 
-func (_ *NitroCertProvider) RootsWithCerts(
-	pemCerts []byte,
-) (
-	*x509.CertPool,
-	error,
-) {
-	certs := x509.NewCertPool()
-	ok := certs.AppendCertsFromPEM(pemCerts)
-	if !ok {
-		return nil, fmt.Errorf("appending cert")
+type FetchingNitroCertProvider struct {
+	HTTPClient   *http.Client
+	RootCerts    *x509.CertPool
+	ExtractRoots ExtractRootsFunc
+}
+
+func NewFetchingNitroCertProvider() *FetchingNitroCertProvider {
```

Comment:
I don't know if it helps, but it may make things simpler.  An alternative way to think about this is that we don't need two CertProviders, instead, we could have just one provider and a way to change the source of the cert bytes (i.e. one object with two builder functions). 

With this design, we start to have a concept in which a cert provider 

1. gets a zipfile (maybe as a io.ReadCloser, e.g. from a file or a http request)
2. read the bytes from the reader
3. check a hash of the bytes
4. extracting certs from the bytes

That decompostion seems to be a little more true to what is going on in the code and may make the testing little simpler.

Reply:
OK, I addressed this by passing in a `ReadCloser` for reading the zip file from embedded variable or the web.
I also integrated the ExtractRoots function into the NitroCertProvider.Roots function. Finally, I removed the `FetchingNitroCertProvider` since all functionality can be in the `NitroCertProvider`